### ISSUE TITLE
Scrambled group labels in pairwise_tukey?

### ISSUE THREAD

**Issue by kncrabtree** [raphaelvallat/pingouin#111]
The group labels in a pairwise_tukey test appear to be mixed up (pingouin version 0.3.6). `men` is a pandas DataFrame with 272 columns. The column `BMXHT` contains numeric data, and the `RIDRETH3` column contains categorical data. Comparing the means in the output to the means given by pandas, it appears that the group labels are scrambled.

![image](https://user-images.githubusercontent.com/20146313/88444431-e9a53a00-cdd1-11ea-8033-50c817ca38fd.png)

I do not observe the same behavior with pairwise_ttests; the T statistics seem to match what I expect from the group labels.

![image](https://user-images.githubusercontent.com/20146313/88444830-0cd0e900-cdd4-11ea-84b7-55d482f6324b.png)

**Comment by raphaelvallat** [comment]
Hi @kncrabtree,

Thanks for opening the issue. This is definitely worrisome. However, I checked the code and the unit tests and I cannot reproduce the error (with pandas = 1.05, numpy = 1.19, pingouin = 0.3.6). Is there any chance you could share your code and dataset (no need to include all 272 columns!). If you forever to keep it private, please feel free to reach out via email at raphaelvallat9@gmail.com

Thanks,
Raphael

**Comment by kncrabtree** [comment]
Sure. My pandas is 1.0.1 and numpy is 1.18.1, so maybe the issue has to do with one of those. I've attached a trimmed down dataframe with only the 2 columns, and minimal code is below (screenshot is my output, which still shows the behavior above). In the meantime, I'll try to update pandas and numpy to see if that resolves the issue.

```
import pandas as pd
import numpy as np
import pingouin

print(f'Pandas: {pd.__version__}')
print(f'Numpy: {np.__version__}')
print(f'Pingouin: {pingouin.__version__}')

df = pd.read_hdf('heights.hd5','df')
pingouin.pairwise_tukey(df,dv='BMXHT',between='RIDRETH3')
df.groupby('RIDRETH3').BMXHT.describe()
```

![image](https://user-images.githubusercontent.com/20146313/88565901-12a01780-cfea-11ea-8e0b-c2172b5b1dfd.png)

[heights.hd5.tar.gz](https://github.com/raphaelvallat/pingouin/files/4983117/heights.hd5.tar.gz)

**Comment by raphaelvallat** [comment]
Hi @kncrabtree,

I just had a look at your data and found the issue. To cut a long story short, this is because RIDRETH3 is encoded as a Categorical. This is more of a Pandas error (or weird behavior at least), but I'm really glad you found it.

The core of the problem is that there is a mismatch in how Pandas and Numpy sort the group labels in the RIDRETH3 column. Specifically, this mismatch occurs because RIDRETH3 is encoded as a categorical column, and not a regular string column. For some weird reasons, it seems that Pandas does not sort the ethnicities in correct alphabetical order but instead uses a custom ordering for categories. Btw, did you define this ordering at any point when creating the categorical column? This custom ordering is then passed to Pingouin, which unfortunately also uses numpy sorting at some point in the code. This mismatch between numpy and pandas sorting is then visible in the output dataframe of the pairwise_tukey function: the labels are based on Numpy sorting, yet all other columns in the dataframe are based on Pandas sorting. Because Pingouin's unit tests for these functions did not include any examples with categorical columns, I never stumbled upon this particular issue. 

You may already have guessed, but a quick fix to this problem is to convert your column to a regular string column:

`df['RIDRETH3'] = df['RIDRETH3'].astype(str)`

However, this has potentially huge implications so I also made changes to the code of Pingouin to get rid of numpy sorting so that the order is only based on Pandas-sorting. For now, the modifications are on the develop branch. If you have the time, I would appreciate if you could double-check that it's now giving correct results.

**Summary**: Pingouin returned wrong group labels in the pairwise_tukey and pairwise_gameshowell function, when, and only when, the between column is encoded as a Categorical. I will make a release very soon to fix this issue.

**Comment by kncrabtree** [comment]
Ah, yes, that makes sense. The original data (part of the [CDC NHANES dataset](https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/DEMO_J.htm#RIDRETH3)) had the column coded as integers, and I converted it into a categorical column for teaching purposes-- the code is part of an example for a new scientific programming in chemistry course I'm developing. So for "real" usage it's pretty unlikely I would have come across this at all. 

I pulled the develop branch and it does indeed give the correct results now. Thanks for resolving this so quickly, and thanks for developing this package! For this class, I was getting a little frustrated as I was trying to demonstrate some very basic statistics. The scipy.stats functions gave very little feedback about the calculations, and the statsmodels API was going to be too difficult to teach to the students in a short time period. This package is just right for combining ease of use with detailed feedback.
